installer/modules.go

```diff
@@ -0,0 +1,939 @@
+
+	// Temporary variables during operation.
+	downloader    *moduleDownload
+	processKiller *delayKiller
+}
+
+type delayKiller struct {
+	proc       *os.Process
+	killer     *time.Timer
+	hardKiller *time.Timer
+}
+
+// kill9After is time after killAfter as expired, not total time. Note that
+// this will kill the process group, so *make sure* the task is created with a
+// new process group or you will probably kill your session.
+func newDelayKiller(proc *os.Process, killAfter, kill9After time.Duration) *delayKiller {
+	k := &delayKiller{
+		proc: proc,
+	}
+	k.killer = time.AfterFunc(killAfter, func() {
+		log.Errorf("Process %d timed out. Sending SIGTERM", k.proc.Pid)
+		// Kill process group (notice minus sign).
+		syscall.Kill(-k.proc.Pid, syscall.SIGTERM)
+	})
+	k.hardKiller = time.AfterFunc(killAfter+kill9After, func() {
+		log.Errorf("Process %d timed out. Sending SIGKILL", k.proc.Pid)
+		// Kill process group (notice minus sign).
+		syscall.Kill(-k.proc.Pid, syscall.SIGKILL)
+	})
+	return k
+}
+
+func (k *delayKiller) Stop() {
+	k.killer.Stop()
+	k.hardKiller.Stop()
+}
+
+func (mod *ModuleInstaller) callModule(state string, capture bool) (string, error) {
+	payloadPath := mod.payloadPath()
+
+	log.Debugf("Calling module: %s %s %s", mod.programPath, state, payloadPath)
+	cmd := exec.Command(mod.programPath, state, payloadPath)
+	cmd.Dir = mod.payloadPath()
+
+	stdoutLogger := newReadLogger(capture)
+	stderrLogger := newReadLogger(false)
+	cmd.Stdout = stdoutLogger
+	cmd.Stderr = stderrLogger
+	// Create new process group so we can kill them all instead of just the parent.
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid: true,
+	}
+
+	err := cmd.Start()
+	if err != nil {
+		log.Errorf("Could not execute update module: %s", err.Error())
+		return "", err
+	}
+
+	timeout := time.Duration(mod.moduleTimeoutSecs) * time.Second
+	// One extra minute to clean up before sending SIGKILL
+	killer := newDelayKiller(cmd.Process, timeout, 1*time.Minute)
+	defer killer.Stop()
+
+	err = cmd.Wait()
+	if err != nil {
+		err = errors.Wrap(err, "Update module terminated abnormally")
+		log.Error(err.Error())
+	}
+
+	return stdoutLogger.output, err
+}
+
+type ReadLogger struct {
+	output  string
+	capture bool
+}
+
+func newReadLogger(capture bool) *ReadLogger {
+	return &ReadLogger{
+		output:  "",
+		capture: capture,
+	}
+}
+
+func (rl *ReadLogger) Write(p []byte) (int, error) {
+
+	buf := string(p)
+	buf = strings.TrimRight(buf, "\n")
+
+	if rl.capture {
+		for _, line := range strings.Split(buf, "\n") {
+			log.Debugf("Update module output: %s", line)
+		}
+		rl.output = rl.output + buf
+	} else {
+		for _, line := range strings.Split(buf, "\n") {
+			log.Infof("Update module output: %s", line)
+		}
+	}
+	return len(p), nil
+}
+
+func (mod *ModuleInstaller) payloadPath() string {
+	index := fmt.Sprintf("%04d", mod.payloadIndex)
+	return path.Join(mod.modulesWorkPath, "payloads", index, "tree")
+}
+
+type fileNameAndContent struct {
```

Comment:
How about `fileProperty` or just `property`?

Reply:
I don't think I agree. The struct doesn't contain file properties (owner, permission mask, etc), it contains content, just like it says.